### DISCUSSION TITLE
Simplified in-line ElementBuilder<T>

### DISCUSSION THREAD

**Discussion by jzabroski** [AutoFixture/AutoFixture#1179] [Ideas]
Frameworks like GenFu and ObjectHydrator have handy ways to create "constrained values".

It would be nice if AutoFixture had ways to simplify the value factory to a `With` expression (added in AutoFixture 4.6.0):

```c#
public static class One
{
  ElementBuilder<T> Of(params object[] elements);
  ElementBuilder<T> Of(IEnumerable<T> elements);
}
```
Example usage:
```c#
public class Transaction {
  public int DaysOutstanding {get; set;}
  public string Code {get; set;}
}
Fixture.Customize<Transaction>(c =>
  c.With(p => p.Code, One.Of(new {"Buy","Sell"})
  c.With(p => p.DaysOutstanding, One.Of(Enumerable.Range(1,30));
```

The advantage is its:
1. Clean API - Just focuses on building elements, easy to read
2. IntelliSense friendly, easy to type
3. DRYs up random sequence generator logic pervasive across AutoFixture plug-ins and extensions

Usually, we want collections built off the following one or more combos:
1. at most one element is specific value, the rest we don't care (pattern matching test)
2. all elements are unique
3. all elements are sub-sets of some other set of elements
4. all elements are unique sub-sets of some other set of elements, etc... 
5. monotonically increasing "sequence"

Some design considerations:
1. The random number generator should belong to the `ElementBuilder<T>`.  This prevents running out of unique strings or objects across parallel tests.  The `ElementBuilder<T>` can then be disposed of when the test completes, and other tests have no knowledge of uniqueness.

**Comment by schalex-dev** [comment]
Is this under any consideration to add to AutoFixture? I would love to use that!

**Reply by aivascu** [reply]
Quite a late reply to this thread but yes this is under consideration. It will be tracked in #1421.
I'm planning the `One.Of()` semantics to be implemented as static ISpecimenBuilder factories.
Since it affects one of the more obscure parts of the code it might take a bit longer for me to prototype this and deliver.
If anyone wishes to help PRs are welcome.
